### PR TITLE
*: rewrite builtin function GREATEST, LEAST

### PR DESCRIPTION
**Pull request by XuHuaiyu**
to #4080 
PTAL @zz-jason @breeswish @jackysp

### CODE REVIEW

File: expression/builtin_cast.go [hunk 1]
```diff
@@ -1348,7 +1348,8 @@ func WrapWithCastAsDecimal(expr Expression, ctx context.Context) Expression {
 // of expr is not type string,
 // otherwise, returns `expr` directly.
 func WrapWithCastAsString(expr Expression, ctx context.Context) Expression {
-	if expr.GetTypeClass() == types.ClassString {
+	exprTp := expr.GetType().Tp
+	if expr.GetTypeClass() == types.ClassString && !types.IsTypeTime(exprTp) && exprTp != mysql.TypeDuration {
```
Comment (shenli) [inline expression/builtin_cast.go]: Is this related with greatest/least?

File: expression/builtin_compare.go [hunk 2]
```diff
@@ -284,98 +291,404 @@ func (b *builtinCoalesceDurationSig) evalDuration(row []types.Datum) (res types.
 	return res, isNull, errors.Trace(err)
 }
 
+// temporalWithDateAsNumTypeClass makes DATE, DATETIME, TIMESTAMP pretend to be numbers rather than strings.
```
Comment (shenli) [inline expression/builtin_compare.go]: Why?

File: util/types/field_type.go [hunk 3]
```diff
@@ -348,7 +348,7 @@ func DefaultTypeForValue(value interface{}, tp *FieldType) {
 		case mysql.TypeDatetime, mysql.TypeTimestamp:
 			tp.Flen = mysql.MaxDatetimeWidthNoFsp
 			if x.Fsp > DefaultFsp { // consider point('.') and the fractional part.
-				tp.Flen = x.Fsp + 1
+				tp.Flen += x.Fsp + 1
```
Comment (shenli) [inline util/types/field_type.go]: Is this a bug fix? Any test covers this?

File: expression/builtin_cast.go (hunk 1 above)
Comment (jackysp) [reply expression/builtin_cast.go]: It is better to add a comment here.

Comment (XuHuaiyu) [reply expression/builtin_cast.go]: It's a bug fix,
if we invoke evalString on a TimeExpr, it would raise an error.

File: expression/builtin_compare.go (hunk 2 above)
Comment (XuHuaiyu) [reply expression/builtin_compare.go]: ported from 
1. https://github.com/mysql/mysql-server/blob/5.7/sql/item.h#L1080
2. https://github.com/mysql/mysql-server/blob/5.7/sql/item_func.cc#L3557

File: expression/builtin_compare.go [hunk 4]
```diff
@@ -284,98 +291,404 @@ func (b *builtinCoalesceDurationSig) evalDuration(row []types.Datum) (res types.
...
+func temporalWithDateAsNumTypeClass(argTp *types.FieldType) (argTypeClass types.TypeClass, isStr bool, isTemporalWithDate bool) {
+	argTypeClass = argTp.ToClass()
```
Comment (zz-jason) [inline expression/builtin_compare.go]: can we use `evalTp` instead of `TypeClass` ?

Comment (XuHuaiyu) [reply expression/builtin_compare.go]: The modification will influence the implementation of `getCmpType`,
thus influence the functions which use it.
Better to change this together.

**Comment by zz-jason** [comment]
LGTM

**Comment by zz-jason** [comment]
@jackysp @breeswish PTAL
